ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala

```diff
@@ -201,6 +227,44 @@ final class ApiParticipantPruningService private (
       loggingContext: LoggingContext
   ): ContextualizedErrorLogger =
     new DamlContextualizedErrorLogger(logger, loggingContext, Some(submissionId))
+
+  override def latestPrunedOffsets(
+      request: LatestPrunedOffsetsRequest
+  ): Future[LatestPrunedOffsetsResponse] =
+    readBackend.lastPrunedOffsets().map { case (divulgencePrunedUpTo, prunedUpToInclusive) =>
+      LatestPrunedOffsetsResponse(
+        participantPrunedUpToInclusive =
+          Some(LedgerOffset(LedgerOffset.Value.Absolute(divulgencePrunedUpTo.value))),
+        allDivulgedContractsPrunedUpToInclusive =
+          Some(LedgerOffset(LedgerOffset.Value.Absolute(prunedUpToInclusive.value))),
+      )
+    }
+}
+
+trait SinglePruningOpEnforcer extends ParticipantPruningServiceGrpc.ParticipantPruningService {
```

Comment:
> I would recommend using composition over inheritance (class or function as opposed to mixin)

sure, why not

> this tool could be made completely independent from ParticipantPruningService, with having some kind of an interface like (=> Future[X], => Throwable) => Future[X]

Hmm, will look into it but didn't consider making it generic since I don't know about other potential use cases. Do you?